src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+        /// Ensures that the priority queue has the specified capacity
+        /// and resizes its underlying array if necessary.
+        /// </summary>
+        public void EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            if (capacity <= this.nodes.Count)
+            {
+                return;
+            }
+
+            this.nodes.Capacity = capacity;
+        }
+
+        /// <summary>
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            this.nodes.TrimExcess();
+        }
+
+        /// <summary>
+        /// Removes the node at the specified index.
+        /// </summary>
+        private void Remove(int indexOfNodeToRemove)
+        {
+            // The idea is to replace the specified node by the very last
+            // node and shorten the array by one.
+
+            var lastNodeIndex = this.GetLastNodeIndex();
+            var lastNode = this.nodes[lastNodeIndex];
+            this.nodes.RemoveAt(lastNodeIndex);
+
+            // In case we wanted to remove the node that was the last one,
+            // we are done.
+
+            if (indexOfNodeToRemove == lastNodeIndex)
+            {
+                return;
+            }
+
+            // Our last node was erased from the array and needs to be
+            // inserted again. Of course, we will overwrite the node we
+            // wanted to remove. After that operation, we will need
+            // to restore the heap property (in general).
+
+            var nodeToRemove = this.nodes[indexOfNodeToRemove];
+
+            var relation = this.Comparer.Compare(lastNode.priority, nodeToRemove.priority);
+            this.PutAt(lastNode, indexOfNodeToRemove);
+
+            if (relation < 0)
+            {
+                this.MoveUp(lastNode, indexOfNodeToRemove);
+            }
+            else
+            {
+                this.MoveDown(lastNode, indexOfNodeToRemove);
+            }
+        }
+
+        /// <summary>
+        /// Puts a node at the specified index.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void PutAt((TElement element, TPriority priority) node, int index)
+        {
+            this.nodes[index] = node;
+        }
+
+        /// <summary>
+        /// Gets the index of the last node in the heap.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetLastNodeIndex() => this.nodes.Count - 1;
+
+        /// <summary>
+        /// Gets the index of an element's parent.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetParentIndex(int index) => (index - 1) / Arity;
```

Comment:
Given you're going with a quaternary heap, you could probably improve this by writing it as
```suggestion
        private int GetParentIndex(int index) => (index - 1) >> 2;
```

Reply:
Hmm, wanted to achieve using a constant in the expression so that it's clear what the number stands for, I guess I could make `Arity` represent a power of `2` then, or something like this.

Anyways, I would assume that the compiler would optimize a constant expression like e.g. `value / 4` into `value >> 2`.

Reply:
> I would assume that the compiler would optimize a constant expression like e.g. value / 4 into value >> 2. Is it not the case?

With `value` being signed, those operations are not always equivalent, so the former is not optimized into the latter, e.g. if `value` was `-1`, they would produce different results.  If `value` were a `uint` instead of an `int`, then yes, the JIT will generate identical code for both `value / 4` and `value >> 2`.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYRoG8aHunGBLAHYYGAZT4BzATAAm5ABSDhg6TAQBKBgF4AfA2WqGAegYoA3Fx709Q0RKnTSCm/vVbdLhtt2lz1HgwtuKwBXRQYAVQFcOxl5UOcBFVcdaySjE19/ELDI6MkZR3ilRNUNFI8vBh8aAF8gA==

Reply:
Thanks, that's very insightful! And the tool you linked is super cool ^^